### PR TITLE
Merge signing & asset manifests into a single manifest

### PR DESCRIPTION
**Pull request by jonfortescue**
Should make there be one manifest for both signing and assets without breaking previous functionality.

Working with @JohnTortugo this afternoon to see about versioning and pushing this up.

### CODE REVIEW

**Comment by chcosta** [comment]
Test failures look related to these changes

**Comment by chcosta** [comment]
Also, do you have a run which includes your changes and works in Arcade?  ie, typical pattern is to produce an official build of your changes, then publish them to the "General Testing" channel, then do another build of Arcade but this time with the published version of Arcade you produced and any yaml changes necessary to show that it's actually working.

**Comment by jonfortescue** [comment]
@chcosta ah, no sorry. will get one for you after i figure out what's causing these test failures.

**Comment by mmitche** [comment]
@jonfortescue I'll try my best to get to this today.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/GenerateBuildManifest.proj [hunk 1]
```diff
@@ -27,8 +27,14 @@
   </ItemGroup>
   <UsingTask TaskName="GenerateBuildManifest" AssemblyFile="$(PkgMicrosoft_DotNet_Build_Tasks_Feed)\tools\netcoreapp2.1\Microsoft.DotNet.Build.Tasks.Feed.dll" />
 
+  <Import Project="../Sign.props" />
+  
   <Target Name="Execute">
     <GenerateBuildManifest Artifacts="@(ItemsToPush)"
+                          ItemsToSign="@(ItemsToSign)"
```
Comment (tmat) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/SdkTasks/GenerateBuildManifest.proj:34]: >  [](start = 25, length = 1)

Nit: indent.

**Comment by mmitche** [comment]
@jonfortescue Do you have a sample of how the signing manifest looks in this case? (and potentially a build of a repo with the new functionality?)

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/BuildManifestUtil.cs [hunk 2]
```diff
@@ -101,7 +105,56 @@ public static void WriteAsXml(this BuildModel buildModel, string filePath, TaskL
...
+            List<FileSignInfoModel> parsedFileSignInfo = new List<FileSignInfoModel>();
+            List<FileExtensionSignInfoModel> parsedFileExtensionSignInfoModel = new List<FileExtensionSignInfoModel>();
+
+            if (itemsToSign != null)
+            {
+                foreach (var itemToSign in itemsToSign)
+                {
+                    var filename = itemToSign.ItemSpec.Replace('\\', '/');
+                    {
+                        var metadata = itemToSign.CloneCustomMetadata() as Dictionary<string, string>;
```
[the remark this answers is not in the document]
Comment (jonfortescue) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/BuildManifestUtil.cs]: it seems it was not used so i removed it!

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/Model/SigningInformationModel.cs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+        public void Add(SigningInformationModel source)
+        {
+            FileExtensionSignInfo.AddRange(source.FileExtensionSignInfo);
+            FileSignInfo.AddRange(source.FileSignInfo);
+            ItemsToSign.AddRange(source.ItemsToSign);
+            StrongNameSignInfo.AddRange(source.StrongNameSignInfo);
+        }
+
+        public XElement ToXml() => IsEmpty() ? null : new XElement(
```
[the remark this answers is not in the document]
Comment (jonfortescue) [reply src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/Model/SigningInformationModel.cs:33]: added!

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/Model/SigningInformationModel.cs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                .Select(f => f.ToXml()))
+            .Concat(ItemsToSign
+                .OrderBy(i => i.File, StringComparer.OrdinalIgnoreCase)
+                .Select(i => i.ToXml()))
+            .Concat(StrongNameSignInfo
+                .OrderBy(s => s.File, StringComparer.OrdinalIgnoreCase)
+                .ThenBy(s => s.PublicKeyToken, StringComparer.OrdinalIgnoreCase)
+                .Select(s => s.ToXml())));
+
+        public bool IsEmpty() => FileExtensionSignInfo.Count == 0 && FileSignInfo.Count == 0
```
[the remark this answers is not in the document]
Comment (jonfortescue) [reply src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/Model/SigningInformationModel.cs]: done; thanks for the tip i like this way of doing things

File: src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/Model/SigningInformationModel.cs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+        public static SigningInformationModel Parse(XElement xml) => xml == null ? null : new SigningInformationModel
+        {
+            FileExtensionSignInfo = xml.Elements("FileExtensionSignInfo").Select(FileExtensionSignInfoModel.Parse).ToList(),
+            FileSignInfo = xml.Elements("FileSignInfo").Select(FileSignInfoModel.Parse).ToList(),
+            ItemsToSign = xml.Elements("ItemsToSign").Select(ItemToSignModel.Parse).ToList(),
+            StrongNameSignInfo = xml.Elements("StrongNameSignInfo").Select(StrongNameSignInfoModel.Parse).ToList(),
+        };
+    }
+    public class FileExtensionSignInfoModel
```
[the remark this answers is not in the document]
Comment (jonfortescue) [reply src/Microsoft.DotNet.VersionTools/lib/src/BuildManifest/Model/SigningInformationModel.cs:64]: done

**Comment by jonfortescue** [comment]
It took me far too long to figure out azure package feeds, but here's a build showing this working (manifest is in the artifacts): https://dnceng.visualstudio.com/internal/_build/results?buildId=712519&view=artifacts&type=publishedArtifacts

File: eng/update-packagesource.ps1 [hunk 6]
```diff
@@ -61,6 +61,8 @@ try {
   AddSourceToNugetConfig $nugetConfigPath $packagesSource
   CheckExitCode "Adding source to NuGet.config" $?
 
+  Get-Content $nugetConfigPath
```
Comment (mmitche) [inline eng/update-packagesource.ps1:64]: Debug logging?

Comment (jonfortescue) [reply eng/update-packagesource.ps1:64]: good catch; thanks!

**Comment by riarenas** [comment]
The new validation for matching manifest attributes is proving to be problematic at least in aspnetcore and installer, and might be the case in other repos as well. The azure devops build information is not populated the same way in every manifest. Wonder if the validation should only happen for non-empty attributes?

See https://github.com/dotnet/core-eng/issues/10262

**Pull request merged**